Doppler.PushContact/Controllers/MessageController.cs

```diff
@@ -0,0 +1,83 @@
+using Microsoft.AspNetCore.Mvc;
+using Doppler.PushContact.DopplerSecurity;
+using Doppler.PushContact.Services;
+using Doppler.PushContact.Services.Messages;
+using Doppler.PushContact.Models;
+
+namespace Doppler.PushContact.Controllers
+{
+    [Authorize(Policies.ONLY_SUPERUSER)]
+    [ApiController]
+    public class MessageController : ControllerBase
+    {
+        private readonly IMessageSender _messageSender;
+        private readonly IMessageRepository _messageRepository;
+        private readonly IPushContactService _pushContactService;
+
+        public MessageController(IPushContactService pushContactService, IMessageRepository messageRepository, IMessageSender messageSender)
+        {
+            _pushContactService = pushContactService;
+            _messageRepository = messageRepository;
+            _messageSender = messageSender;
+        }
+
+        [HttpPost]
+        [Route("message/{messageId}")]
+        public async Task<IActionResult> MessageByVisitorGuid([FromRoute] Guid messageId, [FromBody] string visitorGuid)
+        {
+            if (string.IsNullOrEmpty(visitorGuid) || string.IsNullOrWhiteSpace(visitorGuid))
+            {
+                return BadRequest($"'{nameof(visitorGuid)}' cannot be null, empty or whitespace.");
+            }
+            if (string.IsNullOrEmpty(messageId.ToString()) || string.IsNullOrWhiteSpace(messageId.ToString()))
+            {
+                return BadRequest($"'{nameof(messageId)}' cannot be null, empty or whitespace.");
+            }
+
+            var deviceTokens = await _pushContactService.GetAllDeviceTokensByVisitorGuidAsync(visitorGuid);
+            var message = await _messageRepository.GetMessageDetailsByMessageIdAsync(messageId);
+
+            var sendMessageResult = await _messageSender.SendAsync(message.Title, message.Body, deviceTokens, message.OnClickLinkPropName, message.ImageUrl);
+
+            var notValidTargetDeviceToken = sendMessageResult
+                .SendMessageTargetResult?
+                .Where(x => !x.IsValidTargetDeviceToken)
+                .Select(x => x.TargetDeviceToken);
+
+            if (notValidTargetDeviceToken != null && notValidTargetDeviceToken.Any())
+            {
+                await _pushContactService.DeleteByDeviceTokenAsync(notValidTargetDeviceToken);
+            }
+
+            var now = DateTime.UtcNow;
+
+            var pushContactHistoryEvents = sendMessageResult
+                .SendMessageTargetResult?
+                    .Select(x =>
+                    {
+                        return new PushContactHistoryEvent
+                        {
+                            DeviceToken = x.TargetDeviceToken,
+                            SentSuccess = x.IsSuccess,
+                            EventDate = now,
+                            Details = x.NotSuccessErrorDetails,
+                            MessageId = messageId
+                        };
+                    });
```

Comment:
Very minor, but it is less verbose and, in my opinion, more clear:

```suggestion
                    .Select(x => new PushContactHistoryEvent
                    {
                        DeviceToken = x.TargetDeviceToken,
                        SentSuccess = x.IsSuccess,
                        EventDate = now,
                        Details = x.NotSuccessErrorDetails,
                        MessageId = messageId
                    });
```